### ISSUE TITLE
[Example Request] gRPC authorization with IdentityServer4 (and roles)

### ISSUE THREAD

**Issue by JeepNL** [grpc/grpc-dotnet#949]
For my Blazor (3.2.0) WASM project (.NET 5.0 Preview latest) I'm now using IdentityServer4 (default Visual Studio Template) and gRPC (EF Core/SQLite) data services. 

I would like to protect the gRPC services against unauthorized access much like the [Ticketer ](https://github.com/grpc/grpc-dotnet/tree/master/examples#ticketer)example with `[Authorize]` on the service, and use it with IdentityServer4 roles `[Authorize (Roles=Administrator)]` See: Microsoft Docs, [Secure an ASP.NET Core Blazor WebAssembly hosted app with Identity Server / Name and role claim with API authorization](https://docs.microsoft.com/en-us/aspnet/core/security/blazor/webassembly/hosted-with-identity-server?view=aspnetcore-3.1&tabs=visual-studio#name-and-role-claim-with-api-authorization) 

Maybe it's an idea to add an example to the samples gallery at https://github.com/grpc/grpc-dotnet/tree/master/examples

I tried to figure it out myself, but I'm stuck. I got the two (IdentityServer4 & Ticketer) working separately but I'm stuck combining gRPC authorization with IdentityServer4/Roles.

**Comment by JamesNK** [comment]
Thanks for the suggestion. This isn't a high priority but there is nothing stopping someone contributing a minimal example.

**Comment by JeepNL** [comment]
I understand, and I'm not in a hurry :). I hope someone picks this up. and yes, a minimal example is all what is needed. What should I do with this request, should it remain open or do you want to close it?

**Comment by JamesNK** [comment]
Leave it open.

**Comment by javiercn** [comment]
@JeepNL I wrote a sample for GRPC+Auth that you can find [here](https://github.com/javiercn/BlazorGrpcAuth).

Its lacking roles, but you can add that piece following the docs if that helps you.

**Comment by JeepNL** [comment]
I followed your example and got gRPC authorization working but without `new AuthorizeAttribute { Roles = "Administrator"}` 

As soon as I add that get this error in the client console:"

```
blazor.webassembly.js:1 crit: Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Status(StatusCode="PermissionDenied", Detail="Bad gRPC response. HTTP status code: 403")
Grpc.Core.RpcException: Status(StatusCode="PermissionDenied", Detail="Bad gRPC response. HTTP status code: 403")
  at OpenWiki.Client.Pages.GreeterPage.HandleValidSubmitAsync () [0x0005b] in D:\Repos\OpenWiki\OpenWiki\Client\Pages\GreeterPage.razor:72 
  at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion (System.Threading.Tasks.Task task) <0x30a0ed0 + 0x000da> in <filename unknown>:0 
  at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask (System.Threading.Tasks.Task taskToHandle) <0x30a1a18 + 0x000b6> in <filename unknown>:0 
```

Roles are working, I followed your [other example with Roles](https://github.com/javiercn/BlazorAuthRoles) With the (fake) email address admin[at]ictz.net I have access to the Roles Page. The "user[at]ictz.net" (which is not in 'Administrators', but in the 'Users' Role) doesn't have access to that page. I also tried it with a single "Administrators" role for the admin user. 

Server/**RolesController.cs** ( `[Authorize(Roles = "Administrators")]`)

```
namespace OpenWiki.Server.Controllers
{
    [Authorize(Roles = "Administrators")]
    public class RolesController : ControllerBase
    {
        private readonly UserManager<ApplicationUser> _userManager;
        private readonly RoleManager<IdentityRole> _roleManager;

        public RolesController(
            UserManager<ApplicationUser> userManager,
            RoleManager<IdentityRole> roleManager)
        {
            _userManager = userManager;
            _roleManager = roleManager;
        }

        [HttpPost("User/AddRole")] 
etc... etc...

```

![Annotation 2020-06-19 132852](https://user-images.githubusercontent.com/39458532/85201496-19a77c00-b300-11ea-9584-4f647ba5e165.jpg)

**Comment by JeepNL** [comment]
**FYI:**

I'm creating a GitHub repository (my first ever 😉), it's a Visual Studio Default Blazor WASM gRPC template with IdentityServer4 (VS authentication option "individual user accounts" checked) and with 'Roles' hosted on Kestrel. In short: combining your 2 examples into 1. It will auto seed 2 users (admin & user) and auto create Roles (Administrators & Users) for them. It uses SQLite as DB. It will be CTRL-F5'able without any further configuration.

With that I can reproduce the error. I think I'll have it ready today.

But again: I'm not in a hurry, I really appreciate all of the examples you make available and the time you spend on answering questions, I hope this will make it easier for you to see what I'm doing wrong.

**Comment by JeepNL** [comment]
TL;DR (1) Everything Works!!!

TL;DR (2)
> With that I can reproduce the error. 

Well, no. If I follow your code and instructions the app works with gRPC Role authorization. (I tried to use the 'ProfileService' and that caused the error)

I started a new (*.NET 5 Preview*) project from scratch this morning and I've created my first GitHub Repo. Tt combines your *BlazorAuthRoles* & *BlazorGrpcAuth* examples and adds your snippet `new AuthorizeAttribute { Roles = "Administrator"}` from above: https://github.com/JeepNL/Blazor-WASM-Identity-gRPC

Please don't close this issue yet, because I've (I think) a question about using *ProfileService* but I would like to check that first, but my head is spinning now (8pm UTC+2) 😉 Tomorrow, tomorrow 🎵 

(_oh, man I'm so happy now!_)

**Comment by JeepNL** [comment]
(_continued from yesterday_) Still happy! 

I've added 2 "todo's" to the README.md file at https://github.com/JeepNL/Blazor-WASM-Identity-gRPC, (_Sample project: Blazor WASM, IdentityServer4 & gRPC with roles authorization_)

The first 'Todo' is about adding claims and using them in the client. I've extended the ASP.NET Identity AspNetUsers table with an extra 'CustomClaim' field. This is maybe a bit off topic here, but I've commented on an GitHub issue about that and got an answer from Javier. I've seen several questions about this on GitHub and SO but no-one seems to be able to provide an answer.

If I can add working code to use this to this sample project, maybe it will help more developers.

The second 'Todo' isn't that important, I got the gRPC auth working now, but it's about using claims with 'Profile Service' instead of 'API authorization options' as described  [here](https://docs.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/hosted-with-identity-server?view=aspnetcore-3.1&tabs=visual-studio#configure-identity-server) (_MS Docs_)

**Comment by StefH** [comment]
Related to this question and approach, I'm trying to build this example: **Blazor gRPC-Web with Azure Active Directory Authentication**

**Login to Blazor WASM Client works fine:**
![image](https://user-images.githubusercontent.com/249938/85787847-51615a00-b72c-11ea-907e-821ba0bda68a.png)


**Calling an authenticated gRpc service gives back this RPC error:**

> Status(StatusCode="Internal", Detail="Error starting gRPC call. JSException: TypeError: Failed to fetch", DebugException="WebAssembly.JSException: TypeError: Failed to fetch at System.Net.Http.WebAssemblyHttpHandler.doFetch (System.Threading.Tasks.TaskCompletionSource`1[TResult] tcs, System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) <0x3494528 + 0x00a30> in <filename unknown>:0 at System.Net.Http.WebAssemblyHttpHandler.SendAsync (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) <0x346dd90 + 0x00174> in <filename unknown>:0 at Grpc.Net.Client.Web.GrpcWebHandler.SendAsyncCore (System.Net.Http.HttpRequestMessage request, System.Threading.CancellationToken cancellationToken) [0x00092] in /_/src/Grpc.Net.Client.Web/GrpcWebHandler.cs:137 at Grpc.Net.Client.Internal.GrpcCall`2[TRequest,TResponse].RunCall (System.Net.Http.HttpRequestMessage request, System.Nullable`1[T] timeout) [0x0020c] in /_/src/Grpc.Net.Client/Internal/GrpcCall.cs:477 ")

And the console log shows this error:
> Access to fetch at 'https://login.microsoftonline.com/020b0cf3-d6b2-464e-9b2d-45e124244428/oauth2/authorize?client_id=...' (redirected from 'https://localhost:44375/count.Counter/StartCounter') from origin 'https://localhost:44375' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

Full example projects can be found here:
- https://github.com/StefH/Blazor-gRPC-Web/tree/master/ServerAuthenticated
- https://github.com/StefH/Blazor-gRPC-Web/tree/master/ClientAppAuthenticated

Any advice is appreciated.

**Comment by JamesNK** [comment]
You're making a request to another domain, so you need to setup CORS - https://docs.microsoft.com/en-us/aspnet/core/grpc/browser?view=aspnetcore-3.1#grpc-web-and-cors

**Comment by StefH** [comment]
@JamesNK 
I did update the Server project to use CORS as defined in that help-link, however I still get the same error message in the console.
Can you please take a look at my [Server](https://github.com/StefH/Blazor-gRPC-Web/tree/master/ServerAuthenticated) and [Client](https://github.com/StefH/Blazor-gRPC-Web/tree/master/ClientAppAuthenticated) projects?

**Comment by JeepNL** [comment]
Additional User Claims
Source: https://github.com/JeepNL/Blazor-WASM-Identity-gRPC

![screenshot](https://user-images.githubusercontent.com/39458532/85944387-9f49ae00-b936-11ea-8a2d-20e93ba07759.jpg)

**Comment by StefH** [comment]
@JamesNK 
I'll try to explain the flow how I see it in my example solution, can you please give some advice?

| Caller| Scenario | Outcome
| - | - | - |
| 1. Blazor WASM Client | user logs in | user is correctly logged in via Azure AD
| 2. Blazor WASM Client | user calls GRPC sevice `/count.Counter/StartCounter` on Blazor Server | Because the GRPC Counter service needs authentication, Blazor Server sends 302 redirect with location `https://login.microsoftonline.com/{tenant}/oauth2/authorize?` back
| 3. Blazor WASM Client | tries to call the URL defined in the redirect location | This fails : _Access to fetch at 'https://login.microsoftonline.com/x/oauth2/authorize?client_id=xf&redirect_uri=https%3A%2F%2Flocalhost%3A44375%2Fsignin-oidc&response_type=id_token&scope=openid%20profile&response_mode=form_post&nonce=x&state=x&x-client-SKU=ID_NETSTANDARD2_0&x-client-ver=5.5.0.0' (redirected from 'https://localhost:44375/count.Counter/StartCounter') from origin 'https://localhost:44375' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled._

The questions are:
- why does the the server not use the cookie which is valid ? (Or is this not sent by the client?)
- how to solve the CORS issue on the Blazor WASM Client side?

**Comment by JeepNL** [comment]
I got everything working now so if someone wants to close this issue ...
